### ISSUE TITLE
[Bug]: Generate Dev Container feature generates unbuildable `Dockerfile`

### ISSUE THREAD

**Issue by worldofgeese** [jetpack-io/devbox#627]
**Current Behavior (bug)**
The VS Code Devbox extension's feature to generate Dev Container config files generates an unbuildable `Dockerfile`.

```Dockerfile
FROM alpine:3

# Setting up devbox user
ENV DEVBOX_USER=devbox
RUN adduser -h /home/$DEVBOX_USER -D -s /bin/bash $DEVBOX_USER
RUN addgroup sudo
RUN addgroup $DEVBOX_USER sudo
RUN echo " $DEVBOX_USER      ALL=(ALL:ALL) NOPASSWD: ALL" >> /etc/sudoers

# installing dependencies
RUN apk add --no-cache bash binutils git libstdc++ xz sudo

USER $DEVBOX_USER

# installing devbox
RUN wget --quiet --output-document=/dev/stdout https://get.jetpack.io/devbox | bash -s -- -f
RUN chown -R "${DEVBOX_USER}:${DEVBOX_USER}" /usr/local/bin/devbox

# nix installer script
RUN wget --quiet --output-document=/dev/stdout https://nixos.org/nix/install | sh -s -- --no-daemon
RUN . ~/.nix-profile/etc/profile.d/nix.sh
# updating PATH
ENV PATH="/home/${DEVBOX_USER}/.nix-profile/bin:/home/${DEVBOX_USER}/.devbox/nix/profile/default/bin:${PATH}"

WORKDIR /code
COPY devbox.json devbox.json
RUN devbox shell -- echo "Installing packages"
ENTRYPOINT ["devbox"]
CMD ['shell']
```

```.devcontainer.json
{
  "name": "Devbox Remote Container",
  "build": {
    "dockerfile": "./Dockerfile",
    "context": ".."
  },
  "customizations": {
    "vscode": {
      "settings": {
        "python.defaultInterpreterPath": "/devbox/.devbox/nix/profile/default/bin/python3"
      },
      "extensions": [
        "jetpack-io.devbox",
        "ms-python.python",
        \\ A second copy of the Python extension mysteriously included
        "ms-python.python"
      ]
    }
  },
  "remoteUser": "devbox"
}
```

Logs available: [remoteContainers-2023-02-11T12-10-25.774Z.log](https://github.com/jetpack-io/devbox/files/10713608/remoteContainers-2023-02-11T12-10-25.774Z.log)

The damning line is `Error: mkdir /code/.devbox: permission denied`.

**Expected Behavior (fix)**
Generated Dev Container configuration should build without errors when `Dev Containers: Reopen in container` is invoked from within VS Code.

**Additional context**

> devbox version
0.3.3

```devbox.json
{
  "packages": [
    "python311",
    "nixpacks",
    "pdm",
    "nodePackages.pyright",
    "black",
    "isort",
    "python310Packages.debugpy",
    "glab"
  ],
  "shell": {
    "init_hook": "source conf/set-environment.sh"
  },
  "nixpkgs": {
    "commit": "0c9c1b8fa76fa783c1c52157bab97601f52132eb"
  }
}
```

**Comment by mohsenari** [comment]
Thanks for the detailed and helpful explanation @worldofgeese I tried to recreate the issue. I can see the double mention for `"ms-python.python"`. But even with that I'm able to run in container successfully. Do you have other files in your project that might affect the docker build? Any other details would be helpful for me to recreate this issue.

**Comment by gfriloux** [comment]
Hello, as i am having the same issue, in a different configuration, i guess this does not only involves VS Code's extension, and devcontainer.


My context:

devbox.json
```
{
  "packages": [
    "rsync",
    "git",
    "openssh",
    "curl",
    "shellcheck",
    "just",
    "jq",
    "bc",
    "which",
    "openstackclient",
    "s3cmd",
    "awscli2"
  ],
  "shell": {
    "init_hook": "entrypoint.sh"
  },
  "nixpkgs": {
    "commit": "f80ac848e3d6f0c12c52758c0f25c10c97ca3b62"
  }
}
```

Then i simply type:
- `devbox generate dockerfile`
- `docker build -t myimage .`

I end up with the following message:
```
Step 16/18 : RUN devbox shell -- echo "Installing packages"
 ---> Running in c7cffc98968d
✓ Downloading version 0.3.3... [DONE]
✓ Verifying checksum... [DONE]
✓ Unpacking binary... [DONE]

Error: mkdir /code/.devbox: permission denied
```


The `Dockerfile` created by `devbox`:
```
FROM alpine:3

# Setting up devbox user
ENV DEVBOX_USER=devbox
RUN adduser -h /home/$DEVBOX_USER -D -s /bin/bash $DEVBOX_USER
RUN addgroup sudo
RUN addgroup $DEVBOX_USER sudo
RUN echo " $DEVBOX_USER      ALL=(ALL:ALL) NOPASSWD: ALL" >> /etc/sudoers

# installing dependencies
RUN apk add --no-cache bash binutils git libstdc++ xz sudo

USER $DEVBOX_USER

# installing devbox
RUN wget --quiet --output-document=/dev/stdout https://get.jetpack.io/devbox | bash -s -- -f
RUN chown -R "${DEVBOX_USER}:${DEVBOX_USER}" /usr/local/bin/devbox

# nix installer script
RUN wget --quiet --output-document=/dev/stdout https://nixos.org/nix/install | sh -s -- --no-daemon
RUN . ~/.nix-profile/etc/profile.d/nix.sh
# updating PATH
ENV PATH="/home/${DEVBOX_USER}/.nix-profile/bin:/home/${DEVBOX_USER}/.devbox/nix/profile/default/bin:${PATH}"

WORKDIR /code
COPY devbox.json devbox.json
RUN devbox shell -- echo "Installing packages"
ENTRYPOINT ["devbox"]
CMD ['shell']
```

This is happening with `devbox 0.3.3-1` (`I use Arch BTW`).

**Comment by gfriloux** [comment]
Obviously, we end up running `devbox shell` inside `/code` that is owned by `root`:

```
Step 15/20 : COPY devbox.json devbox.json
 ---> Using cache
 ---> 4e843b07f1f3
Step 16/20 : RUN whoami
 ---> Running in 3663ec57855c
devbox
Removing intermediate container 3663ec57855c
 ---> 47f1bd3b5c7f
Step 17/20 : RUN ls -alh
 ---> Running in 00ce5455c077
total 4K     
drwxr-xr-x    1 root     root          22 Feb 16 14:35 .
drwxr-xr-x    1 root     root         148 Feb 16 14:53 ..
-rw-r--r--    1 root     root         316 Feb 16 14:32 devbox.json
Removing intermediate container 00ce5455c077
 ---> 7468aa53753c
Step 18/20 : RUN devbox shell -- echo "Installing packages"
 ---> Running in 2a01ef0fbcf2
✓ Downloading version 0.3.3... [DONE]
✓ Verifying checksum... [DONE]
✓ Unpacking binary... [DONE]

Error: mkdir /code/.devbox: permission denied
The command '/bin/sh -c devbox shell -- echo "Installing packages"' returned a non-zero code: 1
```

**Comment by mohsenari** [comment]
@gfriloux thanks for the details. I'll try to recreate this. But I think you might be right. This seems like the generated Dockerfile needs to setup correct access for `/code`

**Comment by Lagoja** [comment]
This issue should be resolved with PR #657, now shipped in Devbox 0.4.3.
